File: IKMT.m
close all
clear


filename = ['mesh_inc_var_depth'];

gmsh_to_fem(filename)
nodes = load([filename,'_nodes.txt']);
triangles = load([filename,'_elements.txt']);




I = 0;
fmin = 70.5;
fmax = 79.5;
h = 1;
Nf = (fmax-fmin)/h + 1;
freqs = linspace(fmin, fmax, Nf);

c0 = 1500;
f0 = 75;
lambda_0 = c0/f0;
Dm = 20*lambda_0;

Nx = 201;
Ny = 201;

x_sc = 18*lambda_0;
y_sc = 3.5*lambda_0;
x = linspace(x_sc-4*lambda_0,x_sc+4*lambda_0,Nx);
y = linspace(y_sc-4*lambda_0,y_sc+4*lambda_0,Ny);

[X,Y] = meshgrid(x,y);

Nr = 81;
h = Dm/(Nr-1);
y_a = linspace(h, Dm-h, Nr)';
b_a = y_a(end) - y_a(1);
x_a = 2*lambda_0*ones(Nr,1);
for  i = 1:1
    % Load incident field.
    ui = load(['data/inc_f',num2str(freqs(i)),'.mat']);
    uinc = ui.u;

    % Load total field.
    ut = load(['data/tot_f',num2str(freqs(i)),'.mat']);
    utot = ut.u;

    % Create scattered field.
    usc = utot - uinc;

    Gf = load(['data/green_f',num2str(freqs(i)),'.mat']);
    G = Gf.u;

    omega = 2.*pi*freqs(i);
    NPM = floor((2 * Dm * freqs(i))/c0);
    kn = omega/c0;
    m = 1:NPM;
    VV = sqrt(2/Dm) *sin((pi*y_a*m)/Dm);
    lambdam = (m*pi / Dm).^2;
    betam = sqrt(kn*kn - lambdam);
    if betam(end) <=1e-8
        fprintf('Warning: betam is too small for i = %d\n',i);
    end
    Dbinv = diag(betam);

    A = h*VV'*(VV);

    for m = 1 : length(y_a)
%     figure(2)
% trisurf(triangles,nodes(:,1),nodes(:,2),abs(G(m,:)),linestyle="none");
% shading interp
% % title('TPs','interpreter','latex','FontSize',16)
% colorbar
% axis equal; axis image;
% colormap jet
% view([0 0 90])
% % drawnow;
% % shg
% pause(0.2)
    end


    [Ua,Sa,Va] = svd(A);
    Dvinv = diag(1./diag(Sa));
    SJ = zeros(Nr,NPM);

    for j = 1 : NPM
        for ii = 1 : NPM
            SJ(:,j) = SJ(:,j) + Ua(ii,j)*VV(:,ii);
        end
        % SJ(:,j) = SJ(:,j)/Sa(j,j);
    end
    Shat  = SJ.'*usc*SJ;
    pproj = Dbinv*Ua*Shat*Ua'*Dbinv;
    % pproj = h^2*Dbinv*VV'*usc*VV*Dbinv;
    Gp = h*Dbinv*VV'*G;

    for m = 1 : NPM
        for n = 1 : NPM
            I= I + conj(pproj(m,n))*Gp(m,:).*Gp(n,:);
        end  
    end
% end
% return
figure(2)
trisurf(triangles,nodes(:,1),nodes(:,2),abs(I),linestyle="none");
shading interp
% title('TPs','interpreter','latex','FontSize',16)
colorbar
axis equal; axis image;
colormap jet
view([0 0 90])
% drawnow;
% shg
% imagesc(x,y,abs(reshape(I,Nx,Ny)))
hold on
circle([x_sc,y_sc],20,32,'w');
axis equal
axis image
% colormap jet
colorbar
set(gca,'Ydir','reverse')
drawnow()
shg
pause(0.1)
shg
end

File: IKMT.py
import numpy as np
from scipy.io import loadmat
import matplotlib.pyplot as plt
from configs import *


[X,Y] = np.meshgrid(xg,yg)
I = np.zeros((Nx,Ny),dtype='complex')

for  i in range(Nf):
    
    # Load incident field.
    ui = loadmat('data/inc_f'+ str(frequencies[i]) + '.mat')
    uinc = ui['u']
    # Load total field.
    ut = loadmat('data/tot_f'+ str(frequencies[i]) + '.mat')
    utot = ut['u']
   # Create scattered field.
    usc = utot - uinc

    Gf = loadmat('data/green_f' + str(frequencies[i]) + '.mat')
    G = Gf['u']

    # Dm = Dc

    omega = 2.*np.pi*frequencies[i]
    NPM = np.floor((2 * Dm * frequencies[i])/c0).astype(int)
    kn = omega/c0
    m = np.linspace(1,NPM,NPM)
    VV = np.sqrt(2/Dm) *np.sin(np.pi*np.outer(y_a,m)/Dm)
    lambdam = (m*np.pi / Dm)**2
    betam = np.sqrt(kn*kn - lambdam)
    Dbinv = np.diag(betam)
    A = np.zeros((NPM,NPM))
    for m in range(NPM):
        A[m,m] = (b/Dm) - (1/((m+1)*np.pi))* np.sin(((m+1)*np.pi*b)/Dm)*np.cos((2*(m+1)*np.pi*Dm/2)/Dm)
        for n in range(m+1,NPM):
            A[m,n] = (2/((n-m+1)*np.pi))*np.sin(((n-m+1)*np.pi*b)/(2*Dm))*np.cos(((n-m+1)*np.pi*Dm/2)/Dm) - (2/((n+m+1)*np.pi))*np.sin(((n+m+1)*np.pi*b)/(2*Dm))*np.cos(((n+m+1)*np.pi*Dm/2)/Dm)
    
    A = A + np.transpose(A) - np.diag(np.diag(A))

   


    Ua,Sa,Va = np.linalg.svd(A)
    Sa = Sa/Sa[0]
    Dvinv = np.diag(1/Sa)
    SJ = Dvinv @ Ua @ np.transpose(VV)
    pproj  = np.transpose(VV) @ usc @ VV
    # pproj = Ua @ Shat @ np.transpose(Ua)

    # J = 3
    # U,S,V = np.linalg.svd(pproj)
    # pproj = S[J] * np.reshape(U[:,J],(NPM,1)) @ np.transpose(np.reshape(U[:,J],(NPM,1)))
    for i in range(NPM):
        Gp = h*  Dbinv @ np.transpose(VV) @ np.reshape(G,(Nr,Nx*Ny))

    for m  in range(NPM):
        for n in range(NPM):
            I = I + np.conj(pproj[m,n])*Gp[m,:]*Gp[n,:]

I = I/np.max(abs(I))
print(xg)
print(yg)
plt.imshow(abs(np.reshape(I,(Nx,Ny))), cmap='jet', extent = [xg[0], xg[-1], yg[0], yg[-1]], aspect = 'equal',vmin = 0.4,vmax=1)
plt.colorbar()
plt.gca().add_patch(plt.Circle((x_sc,y_sc),b,fill=False,color='w'))
plt.savefig("image2.eps",format='eps')
plt.show()
